### PR TITLE
Feature/room widgets perms

### LINKED ISSUE #3378: Room Widget Permission UX
Related to https://github.com/vector-im/user-stories/issues/7

Before loading a room widget created by another user, the user should be prompted to give explicit permission.
Permission is given once per widget

image

Use SDK to store and retrieve permission

### PR DESCRIPTION
**Pull request by BillCarsonFr**
### Pull Request Checklist

Fixes #3378 
See https://github.com/vector-im/user-stories/issues/7

Depends on https://github.com/matrix-org/matrix-android-sdk/pull/510

image

image

image

* [ ] Changes has been tested on an Android device or Android emulator with API 16
* [ ] UI change has been tested on both light and dark themes
* [x] Pull request is based on the develop branch
* [x] Pull request updates [CHANGES.rst](https://github.com/vector-im/riot-android/blob/develop/CHANGES.rst)
* [x] Pull request includes screenshots or videos if containing UI changes

### CODE REVIEW

**Comment by BillCarsonFr** [comment]
Updated copy and design
image

image

File: vector/src/main/java/im/vector/activity/VectorAppCompatActivity.kt [hunk 1]
```diff
@@ -45,7 +46,7 @@ import org.matrix.androidsdk.core.Log
 /**
  * Parent class for all Activities in Vector application
  */
-abstract class VectorAppCompatActivity : AppCompatActivity() {
+abstract class VectorAppCompatActivity : BaseMvRxActivity() {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/activity/VectorAppCompatActivity.kt]: You are not using MvRx 1.3? (last version)

File: vector/src/main/java/im/vector/activity/VectorRoomActivity.java [hunk 2]
```diff
@@ -925,7 +924,7 @@ public void onClick(final List<Widget> widgets) {
                     }
 
                     new AlertDialog.Builder(VectorRoomActivity.this)
-                            .setSingleChoiceItems(widgetNames.toArray(CharSequences), 0, new DialogInterface.OnClickListener() {
+                            .setItems(widgetNames.toArray(CharSequences),  new DialogInterface.OnClickListener() {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/activity/VectorRoomActivity.java]: Thanks!

File: vector/src/main/java/im/vector/activity/WidgetActivity.kt [hunk 3]
```diff
@@ -20,355 +20,114 @@ package im.vector.activity
...
+    /* ==========================================================================================
+     * UI EVENTS
+     * ========================================================================================== */
 
-            override fun onUnexpectedError(e: Exception) {
-                onError(e.localizedMessage)
-            }
-        })
+    override fun onBackPressed() {
+        //Hugly i know
```
Comment (bmarty) [inline vector/src/main/java/im/vector/activity/WidgetActivity.kt]: Indeed :)

File: vector/src/main/java/im/vector/activity/WidgetActivity.kt [hunk 4]
```diff
@@ -20,355 +20,114 @@ package im.vector.activity
...
     companion object {
-        private val LOG_TAG = WidgetActivity::class.java.simpleName
 
+        private const val FRAGMENT_TAG_PERMISSION = "FRAGMENT_TAG_PERMISSION"
+        private const val FRAGMENT_TAG_WEBVIEW = "WebView"
         /**
          * The linked widget
          */
-        private const val EXTRA_WIDGET_ID = "EXTRA_WIDGET_ID"
+        const val EXTRA_WIDGET_ID = "EXTRA_WIDGET_ID"
```
Comment (bmarty) [inline vector/src/main/java/im/vector/activity/WidgetActivity.kt]: This should be private

File: vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+        webview.loadUrl("about:blank")
+
+        webview.onPause()
+        webview.removeAllViews()
+
+        // NOTE: This pauses JavaScript execution for ALL WebViews,
+        // do not use if you have other WebViews still alive.
+        // If you create another WebView after calling this,
+        // make sure to call mWebView.resumeTimers().
+        webview.pauseTimers()
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt]: Is it not causing pb with other WebView in the app? Ex with RTC call, Jitsi, VectorWebViewActivity, Captcha, etc.

File: vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+                viewModel.webviewLoadSuccess(url)
+            }
+        }
+    }
+
+
+    override fun invalidate() = withState(viewModel) { state ->
+        //            mWidgetTypeTextView.text = it.widgetName
+        when (state.status) {
+            WidgetState.UNKWNOWN           -> {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt]: There is a bit too much 'N' here :)

File: vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+            R.id.action_widget_open_ext -> if (state.formattedURL.complete) {
+                openUrlInExternalBrowser(requireContext(), state.formattedURL.invoke())
+                return@withState true
+            }
+        }
+        return@withState super.onOptionsItemSelected(item)
+    }
+
+    override fun onPrepareOptionsMenu(menu: Menu?) = withState(viewModel) { state ->
+        menu?.findItem(R.id.action_close)?.let {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt]: No need to use 'let' here

File: vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+                settings.allowFileAccessFromFileURLs = true
+                settings.allowUniversalAccessFromFileURLs = true
+
+                settings.displayZoomControls = false
+            }
+
+            // Permission requests
+            it.webChromeClient = object : WebChromeClient() {
+                override fun onPermissionRequest(request: PermissionRequest) {
+                    Handler(Looper.getMainLooper()).post { request.grant(request.resources) }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt]: We grant permission without asking the user?

File: vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetPermissionViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        //For now put all
+        val infoShared = listOf<Int>(
+                R.string.room_widget_permission_display_name,
+                R.string.room_widget_permission_avatar_url,
+                R.string.room_widget_permission_user_id,
+                R.string.room_widget_permission_theme,
+                R.string.room_widget_permission_widget_id,
+                R.string.room_widget_permission_room_id
+        )
+        setState {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetPermissionViewModel.kt]: This should be done in the initialState and not here

File: vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+        if (TextUtils.equals(widget.widgetId, widget.widgetId)) {
+            if (!widget.isActive) {
+                doFinish()
+            }
+        }
+    }
+
+    init {
+        setState {
+            copy(widgetName = widget.humanName)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetViewModel.kt]: same remark

File: vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+        //If it was added by me, consider it as allowed
+        if (widget.widgetEvent.getSender() == session?.myUserId) {
+            onWidgetAllowed(applicationContext)
+            return
+        }
+
+        val isAllowed = session
+                ?.integrationManager
+                ?.getKnownWidgetPermissions()
+                ?.findLast { it.stateEventId == widget.widgetEvent.eventId }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetViewModel.kt]: Why findLast? (and not first()...)

File: vector/src/main/java/im/vector/widgets/WidgetsManager.java [hunk 12]
```diff
@@ -85,7 +85,7 @@ public String getUIUrl() {
     /**
      * Widget error code
      */
-    public class WidgetError extends MatrixError {
+    public static class WidgetError extends MatrixError {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/widgets/WidgetsManager.java]: I prefer this class not to be static, but defined in its own file

File: vector/src/main/res/layout/bottom_sheet_room_widget_permission.xml [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+            android:textAllCaps="true"
+            android:textColor="@color/vector_error_color" />
+
+        <Button
+            android:id="@+id/bottom_sheet_widget_permission_continue_button"
+            style="@style/VectorButtonStyleFlat"
+            android:layout_marginEnd="@dimen/layout_vertical_margin"
+            android:layout_marginRight="@dimen/layout_vertical_margin"
+            android:minWidth="160dp"
+            android:text="@string/_continue"
```
Comment (bmarty) [inline vector/src/main/res/layout/bottom_sheet_room_widget_permission.xml]: wording: "accept" is maybe better than "continue", no?

**Comment by bmarty** [comment]
Also when accepting T&C, I can still edit the screen (checkbox can be changed) after accepting the &C (i.e. during the request)

File: vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt (hunk 5 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt]: Former code was already pausing/resuming in the WidgetActivity life cycle methods

File: vector/src/main/java/im/vector/activity/VectorAppCompatActivity.kt (hunk 1 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/activity/VectorAppCompatActivity.kt]: Will introduce breaking changes

File: vector/src/main/java/im/vector/widgets/WidgetsManager.java (hunk 12 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/widgets/WidgetsManager.java]: moved

File: vector/src/main/res/layout/bottom_sheet_room_widget_permission.xml (hunk 13 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/bottom_sheet_room_widget_permission.xml]: Continue as per design

File: vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetPermissionViewModel.kt (hunk 9 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetPermissionViewModel.kt]: moved

File: vector/src/main/java/im/vector/activity/WidgetActivity.kt (hunk 3 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/activity/WidgetActivity.kt]: refactored ;)

File: vector/src/main/java/im/vector/activity/WidgetActivity.kt (hunk 4 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/activity/WidgetActivity.kt]: cannot be, it's used by the fragment also

File: vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt (hunk 6 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt]: 👀

File: vector/src/main/java/im/vector/activity/VectorAppCompatActivity.kt (hunk 1 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/activity/VectorAppCompatActivity.kt]: ok

File: vector/src/main/java/im/vector/activity/WidgetActivity.kt (hunk 3 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/activity/WidgetActivity.kt]: thanks

File: vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt (hunk 5 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetFragment.kt]: ok...

File: vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetPermissionViewModel.kt (hunk 9 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/fragments/roomwidgets/RoomWidgetPermissionViewModel.kt]: thx

File: vector/src/main/res/layout/bottom_sheet_room_widget_permission.xml (hunk 13 above)
Comment (bmarty) [reply vector/src/main/res/layout/bottom_sheet_room_widget_permission.xml]: :/

File: vector/src/main/java/im/vector/fragments/roomwidgets/WebviewPermissionUtils.kt [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+        AlertDialog.Builder(activity)
+                .setTitle(title)
+                .setMultiChoiceItems(
+                        request.resources.map { webPermissionToHumanReadable(it, activity) }.toTypedArray()
+                        , null
+                ) { dialog, which, isChecked ->
+                    allowedPermissions[which] = allowedPermissions[which].first to isChecked
+                }
+                .setPositiveButton(R.string.room_widget_resource_grant_permission) { dialog, wich ->
+                    activity.runOnUiThread {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/roomwidgets/WebviewPermissionUtils.kt]: we are already in the UI thread here

File: vector/src/main/java/im/vector/fragments/roomwidgets/WebviewPermissionUtils.kt [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+                        request.grant(allowedPermissions.mapNotNull { perm ->
+                            perm.first.takeIf { perm.second }
+                        }.toTypedArray())
+                    }
+                }
+                .setNegativeButton(R.string.room_widget_resource_decline_permission) { dialog, wich ->
+                    activity.runOnUiThread { request.deny() }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/roomwidgets/WebviewPermissionUtils.kt]: same remark

**Pull request merged**